Add Gaussian quadrature integration for line shape models

vsnever opened the pull request:
This PR fixes #366 and makes the following changes:
- adds `Integrator1D` class for calculating definite integrals of `Function1D`,
- adds `GaussianQuadrature` integrator,
- adds `NumericallyIntegrableLineShapeModel` class for lineshapes that cannot be analytically integrated over a spectral bin,
- replaces half sums in `StarkBroadenedLine.add_line()` with integration over spectral bins.

Mateasek commented on cherab/core/model/lineshape.pyx:
```diff
@@ -352,7 +386,8 @@ cdef class StarkBroadenedLine(LineShapeModel):
     }
 
-    def __init__(self, Line line, double wavelength, Species target_species, Plasma plasma, dict stark_model_coefficients=None):
+    def __init__(self, Line line, double wavelength, Species target_species, Plasma plasma, Integrator1D integrator=None,
+                 dict stark_model_coefficients=None):
```
Please consider putting integrator parameter at the end for backwards compatibility

vsnever replied:
Done.

Mateasek reviewed (changes requested):
Hi @vsnever , thank you for bringing this PR. I really like it and I think it is a great idea. I added some minor comments to the code, but I have one conceptual idea. 

For every lineshape, we have to have the integral value of the power within the bin. For some line shapes there exists an analytical function, some have to be done numerically. Also definite integrals are functions of 2 variables. What if you made `Function2D` base class for the `Integrator1D` and then instead of having `LineShapeModel` and `NumericallyIntegrableLineShapeModel` classes, there could be again only one class, which would take a `Function2D` instance as the integrator parameter?  In the code it does not matter really, if you are calling an analytically defined `Function2D` (created with the function framework) of if you call a numerical integrator, also being a `Function2D`.

This would also need to store the `func` parameter from `integrate` method as a attribute of the `Integrator1D` class, but I think that it is only a minor change.

I hope I managed to express myself clearly. What do you think?

vsnever commented:
Thank you very much for the review, @Mateasek. I've implemented all your suggestions. 

`Integrator1D` now inherits from `Function2D`, which is a good idea. The integrand can be set via the `integrand` property in Python or directly accessed using the `function` attribute in the C-interface.

`NumericallyIntegratedLineShapeModel` is removed and the `integrator` attribute is added to the base `LineShapeModel`. The `integrator` is set to `None` by default except for the `StarkBroadenedLine`, which sets it to `GaussianQuadrature`.

vsnever commented on cherab/core/math/integrators/integrators1d.pyx:
```diff
@@ -0,0 +1,199 @@
+
+        i = 0
+        for order in range(self._min_order, self._max_order + 1):
+            self._roots[i:i + order], self._weights[i:i + order] = roots_legendre(order)
```
I rewrote the integration, so `x` now taken as `x = c + d * self._roots_mv[i]`, where `c = 0.5 * (a + b)` and `d = 0.5 * (b - a)`. Adding 1 is no longer required.

vsnever commented on cherab/core/math/integrators/integrators1d.pyx:
```diff
@@ -0,0 +1,199 @@
+                x = a + 0.5 * (b - a) * (self._roots_mv[i] + 1.)
+                newval += self._weights_mv[i] * func.evaluate(x)
+            newval *= 0.5 * (b - a)
+
```
Done.

vsnever commented on cherab/core/math/integrators/integrators1d.pyx:
```diff
@@ -0,0 +1,199 @@
+    Compute a definite integral of a one-dimensional function.
+    """
+
+    cpdef (double, double) integrate(self, Function1D func, double a, double b):
```
The `Integrator1D` now inherits from the `Function2D`. I renamed `integrate` to `evaluate` and now this method returns a single value. Now the `__call__` method from the parent class can be used.

Mateasek reviewed (approved):
Hi @vsnever , thanks for the changes. I think now the code is ready to be accepted.

jacklovell reviewed (changes requested):
I like this a lot: having more specialised integration routines to supplement trapezoidal integration is very useful indeed.

However, I don't think it makes sense to subclass the function framework classes for these integrators: just because `Integrator1D.evaluate` is a function of 2 arguments doesn't make the integrator a 2D function. For example, doing `Integrator1D(function) + Arg1D()` makes no sense.

Cherab has precedent for this: the core rate objects don't subclass from `Function2D`, even though they have `evaluate(ne, te)` methods. I think the same should apply here.

Otherwise, looks good. Couple of minor comments addressed inline.

Mateasek commented on cherab/core/math/integrators/integrators1d.pyx:
```diff
@@ -0,0 +1,213 @@
+            int order, i, ibegin
+            double newval, oldval, error, x, c, d
+
+        if self.function is None:
+            raise AttributeError("Integrand is not set.")
```
I would keep the function as a property. The integration is going to be called many times during observations

vsnever replied:
The performance of the Cython code will not suffer because `evaluate` is called directly in Cython, and the wrapping is done only in the Python `__call__` method. But never mind, let's leave the integrand as a property.

The problem is that we can't make it mandatory to set the integrand on initialisation. This is so because an instance of `Integrator1D` is passed to the `LineShapeModel` constructor as an argument, but the integrand cannot be set prior `LineShapeModel` initialisation. If setting the integrand to `None` is a bad option, then it should have some default value, for example `Constant1D(0)`.

vsnever commented on cherab/core/math/integrators/integrators1d.pxd:
```diff
@@ -0,0 +1,37 @@
+        ndarray _roots, _weights
+        double[:] _roots_mv, _weights_mv
+
+    cdef _build_cash(self)
```
Freudian slip :-). Fixed.

vsnever commented on cherab/core/math/integrators/integrators1d.pyx:
```diff
@@ -0,0 +1,213 @@
+cimport cython
+
+
+cdef class Integrator1D(Function2D):
```
I removed the `Function2D` inheritance and added the `__call__` method.

vsnever commented:
Thanks for the review, @jacklovell. I fixed the issues you mentioned.

Also, I've added analytical line shape normalisation to `StarkBroadenedLine` using the expression:
$$\int_{-Ra}^{Ra}\frac{1}{x^{5/2}+a^{5/2}}dx=\frac{2R}{a^{3/2}}{_2}F_1(\frac{2}{5},1,\frac{7}{5},-R^{5/2}),$$
where $_2F_1$ is a hypergeometric function.
